File: output/xCodeEval_8000/C++/Python/1f6c549a9d407c6276f7b0cacba3bb49.py
mod = 998244353
K = 18

class Mint:
    def __init__(self, n=0):
        self.n = n % mod

    def __neg__(self):
        return Mint(-self.n)

    def __pos__(self):
        return self

    def __iadd__(self, other):
        self.n += other.n
        if self.n >= mod:
            self.n -= mod
        return self

    def __isub__(self, other):
        self.n -= other.n
        if self.n < 0:
            self.n += mod
        return self

    def __imul__(self, other):
        self.n = (self.n * other.n) % mod
        return self

    def __ifloordiv__(self, other):
        self.n = (self.n * other.inv().n) % mod
        return self

    def __eq__(self, other):
        return self.n == other.n

    def __ne__(self, other):
        return self.n != other.n

    def __sub__(self, other):
        return Mint(self.n) - other

    def __add__(self, other):
        return Mint(self.n) + other

    def __mul__(self, other):
        return Mint(self.n) * other

    def __floordiv__(self, other):
        return Mint(self.n) // other

    def pow(self, n):
        res = Mint(1)
        x = Mint(self.n)
        while n > 0:
            if n & 1:
                res *= x
            x *= x
            n >>= 1
        return res

    def inv(self):
        return self.pow(mod - 2)

class Poly:
    def __init__(self, n=-1):
        self.c = [Mint(0)] * (n + 1)

    def deg(self):
        return len(self.c) - 1

    def at(self, n):
        if n < len(self.c):
            return self.c[n]
        else:
            return Mint(0)

    def __iadd__(self, other):
        if other.deg() > self.deg():
            d = list(other.c)
            for i in range(self.deg() + 1):
                d[i] += self.c[i]
            self.c = d
        else:
            for i in range(other.deg() + 1):
                self.c[i] += other.c[i]
        return self

    def __isub__(self, other):
        if other.deg() > self.deg():
            d = [-x for x in other.c]
            for i in range(self.deg() + 1):
                d[i] += self.c[i]
            self.c = d
        else:
            for i in range(other.deg() + 1):
                self.c[i] -= other.c[i]
        return self

    def __imul__(self, other):
        if self.deg() < 0 or other.deg() < 0:
            self.c = []
            return self
        rdeg = self.deg() + other.deg()
        k = 32 - (rdeg).bit_length()
        assert k <= K + 1
        d = [Mint(0)] * (1 << k)
        e = [Mint(0)] * (1 << k)
        for i in range(self.deg() + 1):
            d[i] = self.c[i]
        for i in range(other.deg() + 1):
            e[i] = other.c[i]
        fft(d, k)
        fft(e, k)
        for i in range(1 << k):
            d[i] *= e[i]
        ifft(d, k)
        self.c = d[:rdeg + 1]
        im = Mint((mod + 1) // 2).pow(k)
        for i in range(rdeg + 1):
            self.c[i] *= im
        return self

    def __add__(self, other):
        res = Poly()
        res.c = list(self.c)
        res += other
        return res

    def __sub__(self, other):
        res = Poly()
        res.c = list(self.c)
        res -= other
        return res

    def __mul__(self, other):
        res = Poly()
        res.c = list(self.c)
        res *= other
        return res

    def __floordiv__(self, other):
        res = Poly()
        res.c = list(self.c)
        res //= other
        return res

    def inv(self, n=-1):
        assert self.c[0] != 0
        if n == -1:
            n = self.deg()
        if n == -1:
            return self
        k = 32 - (n).bit_length()
        assert k <= K
        d = [Mint(0)] * (1 << k)
        ee = [Mint(0)] * (1 << k)
        res = Poly(n)
        res.c[0] = self.c[0].inv()
        y = (mod + 1) // 2
        if y & 1:
            y += mod
        y //= 2
        im = Mint(mod - y)
        for i in range(k):
            for j in range(1 << i):
                ff = res.c[j]
                gg = self.c[j << (k - i)]
                d[j] = ff
                ee[j] = gg
                d[j + (1 << i)] = 0
                ee[j + (1 << i)] = 0
            fft(d, i + 1)
            fft(ee, i + 1)
            for j in range(1 << (i + 1)):
                d[j] *= ee[j]
            ifft(d, i + 1)
            for j in range(1 << i):
                d[j] = d[j + (1 << i)]
                d[j + (1 << i)] = Mint(0)
            fft(d, i + 1)
            for j in range(1 << (i + 1)):
                d[j] *= ee[j]
            ifft(d, i + 1)
            for j in range(1 << i, min(1 << (i + 1), n + 1)):
                res.c[j] = d[j] * im
            if im.n & 1:
                im.n += mod
            im.n //= 2
            if im.n & 1:
                im.n += mod
            im.n //= 2
        return res

    def sqrt(self, n=-1):
        if n == -1:
            n = self.deg()
        if n == -1:
            return self
        assert self.c[0] == 1
        if n == 0:
            res = Poly(0)
            res.c[0] = 1
            return res
        k = 32 - (n).bit_length()
        assert k <= K
        g = Poly((1 << k) - 1)
        d = [Mint(0)] * (1 << (k + 1))
        ee = [Mint(0)] * (1 << (k + 1))
        res = Poly(n)
        res.c[0] = 1
        y = (mod + 1) // 2
        if y & 1:
            y += mod
        y //= 2
        if y & 1:
            y += mod
        y //= 2
        im = Mint(mod - y)
        for i in range(k):
            for j in range(1 << (k - i)):
                t = (1 << (k - i)) + j
                g.c = [self.c[(1 << k) - i - l - 1] for l in range(1 << i)]
                gg = g.c
                d[j] = res.c[j]
                ee[j] = gg[j]
                d[j + (1 << i)] = 0
                ee[j + (1 << i)] = 0
            fft(d, i + 2)
            fft(ee, i + 2)
            for j in range(1 << (i + 2)):
                d[j] *= ee[j]
            ifft(d, i + 2)
            for j in range(1 << (i + 1)):
                d[j] = d[j + (1 << (i + 1))]
                d[j + (1 << (i + 1))] = Mint(0)
            fft(d, i + 2)
            for j in range(1 << (i + 2)):
                d[j] *= ee[j]
            ifft(d, i + 2)
            for j in range(1 << (i + 1)):
                res.c[j + (1 << i)] = d[j] * im
            if im.n & 1:
                im.n += mod
            im.n //= 2
        return res

    def exp(self, n=-1):
        if n == -1:
            n = self.deg()
        if n == -1:
            return self
        assert self.c[0] == 0
        if n == 0:
            res = Poly(0)
            res.c[0] = 1
            return res
        k = 32 - (n).bit_length()
        assert k <= K
        A = Poly(1 << k)
        A.c[0] = 1
        A.c[1] = (k - 2) % mod
        A.c[2] = 1
        B = A.log()
        for i in range(len(B.c)):
            B.c[i] *= Mint(n)
        C = B.exp()
        ans = Mint(1).pow(n)
        for i in range(len(C.c)):
            ans -= C.c[i]
        ans *= Mint(n).pow(n)
        return ans

    def log(self, n=-1):
        if n == -1:
            n = self.deg()
        if n == -1:
            return self
        assert self.c[0] == 1
        k = 32 - (n).bit_length()
        assert k <= K
        B = Poly()
        B.c = list(self.c)
        B = B.inv(n)
        dQ = B.deg()
        while dQ >= 0 and B.c[dQ] == 0:
            dQ -= 1
        assert B.c[dQ] != 0
        if dQ > n:
            return Poly()
        P = Poly(min(dQ, n - dQ))
        for j in range(min(dQ, n - dQ) + 1):
            P.c[j] = B.c[dQ - j]
        k = 32 - (n - dQ).bit_length()
        assert k <= K + 1
        d = [Mint(0)] * (1 << k)
        e = [Mint(0)] * (1 << k)
        for i in range(P.deg() + 1):
            d[i] = P.c[i]
        fft(d, k)
        for i in range(Q.deg() + 1):
            e[i] = Q.c[i]
        fft(e, k)
        for i in range(1 << k):
            d[i] *= e[i]
        ifft(d, k)
        im = Mint((mod + 1) // 2).pow(k)
        dR = -1
        for i in range(Q.deg()):
            d[i] = self.c[i] - d[i] * im
            if d[i] != 0:
                dR = i
        self.c = d[:dR + 1]
        return self

    def quotient(self, Q):
        assert Q.deg() >= 0
        if self.deg() < 0:
            return self
        dP = self.deg()
        while dP >= 0 and self.c[dP] == 0:
            dP -= 1
        if self.c[dP] == 0:
            return Poly()
        dQ = Q.deg()
        while dQ >= 0 and Q.c[dQ] == 0:
            dQ -= 1
        assert Q.c[dQ] != 0
        if dQ > dP:
            return Poly()
        RQ = Poly(min(dQ, dP - dQ))
        for j in range(min(dQ, dP - dQ) + 1):
            RQ.c[j] = Q.c[dQ - j]
        k = 32 - (dP - dQ).bit_length()
        assert k <= K + 1
        d = [Mint(0)] * (1 << k)
        e = [Mint(0)] * (1 << k)
        for i in range(dP - dQ + 1):
            d[i] = self.c[dP - i]
        RQ = RQ.inv(dP - dQ)
        for i in range(RQ.deg() + 1):
            e[i] = RQ.c[i]
        fft(d, k)
        fft(e, k)
        for i in range(1 << k):
            d[i] *= e[i]
        ifft(d, k)
        im = Mint((mod + 1) // 2).pow(k)
        Q.c = d[:dP - dQ + 1]
        return self

    def eval(self, x):
        res = Mint(0)
        for i in range(self.deg(), -1, -1):
            res *= x
            res += self.c[i]
        return res

    def eval2(self, P):
        m = len(P)
        if m <= (1 << _ls):
            v = [self.eval(x) for x in P]
            return v
        k = 32 - (m - 1).bit_length()
        v = [Mint(0)] * m
        w = [Mint(0)] * m
        for i in range(m):
            v[i] = self.eval(P[i])
        fft(v, k)
        for i in range(m):
            w[i] = v[i]
        return w

    def interpolation(self, a, b):
        m = len(a)
        assert len(b) == m
        return Poly()

def fft(a, k):
    u = 1
    v = 1 << (k - 1)
    for i in range(k, 0, -1):
        for jh in range(u):
            wj = Mint(3).pow(jh)
            for j in range(jh << i, jh << i | v):
                Ajv = wj * a[j + v]
                a[j + v] = a[j] - Ajv
                a[j] += Ajv
        u <<= 1
        v >>= 1

def ifft(a, k):
    u = 1 << (k - 1)
    v = 1
    for i in range(1, k + 1):
        for jh in range(u):
            wj = Mint(332748118).pow(jh)
            for j in range(jh << i, jh << i | v):
                Ajv = a[j] - a[j + v]
                a[j] += a[j + v]
                a[j + v] = wj * Ajv
        u >>= 1
        v <<= 1

def middle_product(n, lhs, rhs):
    dl = lhs.deg()
    while lhs.c[dl] == 0:
        dl -= 1
    assert rhs.deg() <= n + dl
    k = 32 - (dl + n).bit_length()
    P = Poly((1 << k) - 1)
    Q = Poly((1 << k) - 1)
    for i in range(dl + 1):
        P.c[i] = lhs.c[dl - i]
    for i in range(rhs.deg() + 1):
        Q.c[i] = rhs.c[i]
    fft(P.c, k)
    fft(Q.c, k)
    for i in range(1 << k):
        P.c[i] *= Q.c[i]
    ifft(P.c, k)
    im = Mint((mod + 1) // 2).pow(k)
    Q.c = P.c[:n + 1]
    return Q

def uptree(C, T):
    m = len(C)
    k = 32 - (m - 1).bit_length()
    C = [Poly([C[i]]) for i in range(m)]
    for i in range(k, 0, -1):
        for j in range(1 << (k - i)):
            C[j] = T[(1 << (k - i)) + 2 * j + 1] * C[2 * j] + T[(1 << (k - i)) + 2 * j] * C[2 * j + 1]
    return C[0]

def tuptree(B, T):
    m = B.deg() + 1
    k = 32 - (m - 1).bit_length()
    C = [Poly([B.c[i]]) for i in range(m)]
    for i in range(k, 0, -1):
        for j in range((1 << (k - i)) - 1, -1, -1):
            n = (1 << (i - 1)) - 1
            C[2 * j + 1] = middle_product(n, T[(1 << (k - i + 1)) + 2 * j], C[j])
            C[2 * j] = middle_product(n, T[(1 << (k - i + 1)) + 2 * j + 1], C[j])
    return [C[i].c[0] for i in range(m)]

def solve():
    n, k = map(int, input().split())
    hs = list(map(int, input().split()))

    cnt = 0
    for i in range(n):
        if hs[i] == hs[(i + 1) % n]:
            cnt += 1

    if cnt == n:
        print(0)
        return

    A = Poly(2 * 10**5 + 1000)
    A.c[0] = 1
    A.c[1] = (k - 2) % mod
    A.c[2] = 1

    B = A.log()
    for i in range(len(B.c)):
        B.c[i] *= Mint(n - cnt)

    C = B.exp()

    ans = Mint(k).pow(n - cnt)
    for i in range(n - cnt + 1):
        ans -= C.c[i]

    ans *= Mint(k).pow(cnt)
    print(ans.n)

solve()

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/1f6c549a9d407c6276f7b0cacba3bb49.go
package main

import (
	"fmt"
)

const mod = 998244353
const K = 18

type Mint struct {
	n int
}

func NewMint(n int) Mint {
	return Mint{n: n}
}

func (m Mint) Negate() Mint {
	if m.n != 0 {
		return Mint{n: mod - m.n}
	}
	return Mint{n: 0}
}

func (m Mint) Increment() Mint {
	m.n++
	if m.n == mod {
		m.n = 0
	}
	return m
}

func (m Mint) Decrement() Mint {
	m.n--
	if m.n < 0 {
		m.n = mod - 1
	}
	return m
}

func (m *Mint) Subtract(rhs Mint) {
	if m.n >= rhs.n {
		m.n -= rhs.n
	} else {
		m.n = mod - (rhs.n - m.n)
	}
}

func (m *Mint) Add(rhs Mint) {
	m.n += rhs.n
	if m.n >= mod {
		m.n -= mod
	}
}

func (m *Mint) Multiply(rhs Mint) {
	m.n = (m.n * rhs.n) % mod
}

func (m *Mint) Divide(rhs Mint) {
	m.n = (m.n * rhs.Inverse().n) % mod
}

func (m Mint) Equal(rhs Mint) bool {
	return m.n == rhs.n
}

func (m Mint) NotEqual(rhs Mint) bool {
	return m.n != rhs.n
}

func (m Mint) Pow(n int) Mint {
	r := NewMint(1)
	x := m
	for n > 0 {
		if n&1 == 1 {
			r.Multiply(x)
		}
		x.Multiply(x)
		n >>= 1
	}
	return r
}

func (m Mint) Inverse() Mint {
	return m.Pow(mod - 2)
}

type Poly struct {
	c []Mint
}

func NewPoly(n int) Poly {
	return Poly{c: make([]Mint, n+1)}
}

func (p Poly) Degree() int {
	return len(p.c) - 1
}

func (p Poly) At(n int) Mint {
	if n < len(p.c) {
		return p.c[n]
	}
	return NewMint(0)
}

func (p *Poly) Add(rhs Poly) {
	if rhs.Degree() > p.Degree() {
		d := make([]Mint, rhs.Degree()+1)
		for i := 0; i <= p.Degree(); i++ {
			d[i] = p.c[i]
		}
		p.c = d
	} else if rhs.Degree() < p.Degree() {
		for i := 0; i <= rhs.Degree(); i++ {
			p.c[i].Add(rhs.c[i])
		}
	} else {
		for i := 0; i <= p.Degree(); i++ {
			p.c[i].Add(rhs.c[i])
		}
	}
}

func (p *Poly) Subtract(rhs Poly) {
	if rhs.Degree() > p.Degree() {
		d := make([]Mint, rhs.Degree()+1)
		for i := 0; i <= p.Degree(); i++ {
			d[i] = p.c[i]
		}
		p.c = d
	} else if rhs.Degree() < p.Degree() {
		for i := 0; i <= rhs.Degree(); i++ {
			p.c[i].Subtract(rhs.c[i])
		}
	} else {
		for i := 0; i <= p.Degree(); i++ {
			p.c[i].Subtract(rhs.c[i])
		}
	}
}

func (p *Poly) Multiply(rhs Poly) {
	if p.Degree() < 0 || rhs.Degree() < 0 {
		p.c = make([]Mint, 0)
		return
	}
	rdeg := p.Degree() + rhs.Degree()
	k := 32 - clz(rdeg)
	d := make([]Mint, 1<<k)
	e := make([]Mint, 1<<k)
	for i := 0; i <= p.Degree(); i++ {
		d[i] = p.c[i]
	}
	for i := 0; i <= rhs.Degree(); i++ {
		e[i] = rhs.c[i]
	}
	fft(d, k)
	fft(e, k)
	for i := 0; i < 1<<k; i++ {
		d[i].Multiply(e[i])
	}
	ifft(d, k)
	p.c = make([]Mint, rdeg+1)
	im := NewMint((mod + 1) / 2).Pow(k)
	for i := 0; i <= rdeg; i++ {
		p.c[i] = d[i].Multiply(im)
	}
}

func (p Poly) AddPoly(rhs Poly) Poly {
	r := p
	r.Add(rhs)
	return r
}

func (p Poly) SubtractPoly(rhs Poly) Poly {
	r := p
	r.Subtract(rhs)
	return r
}

func (p Poly) MultiplyPoly(rhs Poly) Poly {
	r := p
	r.Multiply(rhs)
	return r
}

func (p Poly) ModPoly(rhs Poly) Poly {
	r := p
	r.Mod(rhs)
	return r
}

func (p *Poly) Mod(rhs Poly) {
	if rhs.Degree() > p.Degree() {
		return
	}
	if rhs.Degree() < 32 {
		dQ := rhs.Degree()
		for dQ >= 0 && rhs.c[dQ].Equal(NewMint(0)) {
			dQ--
		}
		if rhs.c[dQ].Equal(NewMint(0)) {
			p.c = make([]Mint, 0)
			return
		}
		for i := p.Degree(); i >= dQ; i-- {
			if p.c[i].Equal(NewMint(0)) {
				continue
			}
			x := p.c[i].Divide(rhs.c[dQ])
			p.c[i] = NewMint(0)
			for j := 1; j <= dQ; j++ {
				p.c[i-j].Subtract(x.Multiply(rhs.c[dQ-j]))
			}
		}
		dP := dQ - 1
		for dP >= 0 && p.c[dP].Equal(NewMint(0)) {
			dP--
		}
		p.c = p.c[:dP+1]
		return
	}
	P := p.Quotient(rhs)
	k := 32 - clz(p.Degree())
	v := (1 << k) - 1
	A := middleProduct(v, B, *p)
	B.c = B.c[:v+1]
	for i := 0; i <= v; i++ {
		B.c[i] = A.c[v-i]
	}
}

func (p Poly) Log() Poly {
	n := p.Degree()
	if n == -1 {
		return p
	}
	k := 32 - clz(n)
	B := p
	for i := 0; i < len(B.c); i++ {
		B.c[i] = B.c[i].Multiply(NewMint(n))
	}
	B.c = B.c[:n+1]
	B.c = B.c[1:]
	B.c = append(B.c, NewMint(0))
	B = B.Inverse(n)
	v := (1 << k) - 1
	A := middleProduct(v, B, p)
	B.c = B.c[:v+1]
	for i := 0; i <= v; i++ {
		B.c[i] = A.c[v-i]
	}
	return B
}

func (p Poly) Exp() Poly {
	n := p.Degree()
	if n == -1 {
		return p
	}
	k := 32 - clz(n)
	B := NewPoly(1 << k)
	B.c[0] = NewMint(1)
	B.c[1] = NewMint(n)
	B.c[2] = NewMint(1)
	B = B.Log()
	for i := 0; i < len(B.c); i++ {
		B.c[i] = B.c[i].Multiply(NewMint(n))
	}
	B.c = B.c[:n+1]
	B = B.Exp()
	ans := NewMint(n).Pow(n)
	for i := 0; i <= n; i++ {
		ans = ans.Subtract(B.c[i])
	}
	ans = ans.Multiply(NewMint(n).Pow(n))
	return ans
}

func (p Poly) Quotient(Q Poly) Poly {
	if Q.Degree() < 0 {
		return p
	}
	if p.Degree() < 0 {
		return p
	}

	dP := p.Degree()
	for dP >= 0 && p.c[dP].Equal(NewMint(0)) {
		dP--
	}
	if p.c[dP].Equal(NewMint(0)) {
		return NewPoly(0)
	}

	dQ := Q.Degree()
	for dQ >= 0 && Q.c[dQ].Equal(NewMint(0)) {
		dQ--
	}
	if Q.c[dQ].Equal(NewMint(0)) {
		return NewPoly(0)
	}

	if dQ > dP {
		return NewPoly(0)
	}

	RQ := NewPoly(min(dQ, dP-dQ))
	for j := 0; j <= min(dQ, dP-dQ); j++ {
		RQ.c[j] = Q.c[dQ-j]
	}

	k := 32 - clz(dP - dQ)
	v := (1 << k) - 1
	A := middleProduct(v, RQ, p)
	B := A.Inverse(dP - dQ)
	A = middleProduct(v, B, *p)
	B.c = B.c[:v+1]
	for i := 0; i <= v; i++ {
		B.c[i] = A.c[v-i]
	}

	return B
}

func (p *Poly) Modulo(rhs Poly) {
	if rhs.Degree() > p.Degree() {
		return
	}
	if rhs.Degree() < 32 {
		dQ := rhs.Degree()
		for dQ >= 0 && rhs.c[dQ].Equal(NewMint(0)) {
			dQ--
		}
		if rhs.c[dQ].Equal(NewMint(0)) {
			p.c = make([]Mint, 0)
			return
		}
		for i := p.Degree(); i >= dQ; i-- {
			if p.c[i].Equal(NewMint(0)) {
				continue
			}
			x := p.c[i].Divide(rhs.c[dQ])
			p.c[i] = NewMint(0)
			for j := 1; j <= dQ; j++ {
				p.c[i-j].Subtract(x.Multiply(rhs.c[dQ-j]))
			}
		}
		dP := dQ - 1
		for dP >= 0 && p.c[dP].Equal(NewMint(0)) {
			dP--
		}
		p.c = p.c[:dP+1]
		return
	}
	P := p.Quotient(rhs)
	k := 32 - clz(p.Degree())
	v := (1 << k) - 1
	A := middleProduct(v, B, *p)
	B.c = B.c[:v+1]
	for i := 0; i <= v; i++ {
		B.c[i] = A.c[v-i]
	}
}

func (p Poly) Inverse(n int) Poly {
	if p.c[0].Equal(NewMint(0)) {
		return p
	}
	if n == -1 {
		n = p.Degree()
	}
	if n == -1 {
		return p
	}
	k := 32 - clz(n)
	d := make([]Mint, 1<<k)
	ee := make([]Mint, 1<<k)
	r := NewPoly(n)
	r.c[0] = p.c[0].Inverse()
	y := (mod + 1) / 2
	if y&1 == 1 {
		y += mod
	}
	y /= 2
	im := NewMint(mod - y)
	for i := 0; i < k; i++ {
		for j := 0; j < 1<<(i+1); j++ {
			ee[j] = r.c[j]
		}
		fft(ee, i+1)
		for j := 0; j < 1<<i; j++ {
			d[j] = r.c[j]
			d[j] = d[j].Multiply(im)
		}
		fft(d, i+1)
		for j := 0; j < 1<<(i+1); j++ {
			d[j] = d[j].Multiply(ee[j])
		}
		ifft(d, i+1)
		for j := 1 << i; j < min(1<<(i+1), n+1); j++ {
			r.c[j] = d[j].Multiply(im)
		}
		if im.n&1 == 1 {
			im.n += mod
		}
		im.n /= 2
		if im.n&1 == 1 {
			im.n += mod
		}
		im.n /= 2
	}
	return r
}

func (p Poly) Isqrt(n int) Poly {
	if p.c[0].Equal(NewMint(1)) {
		return p
	}
	if n == -1 {
		n = p.Degree()
	}
	if n == -1 {
		return p
	}
	k := 32 - clz(n)
	r := NewPoly(n)
	r.c[0] = NewMint(1)
	y := (mod + 1) / 2
	if y&1 == 1 {
		y += mod
	}
	y /= 2
	if y&1 == 1 {
		y += mod
	}
	y /= 2
	im := NewMint(mod - y)
	for i := 0; i < k; i++ {
		for j := 0; j < 1<<(i+1); j++ {
			r.c[j] = p.c[j]
		}
		fft(r.c, i+2)
		for j := 0; j < 1<<(i+1); j++ {
			r.c[j] = r.c[j].Multiply(im)
		}
		fft(r.c, i+2)
		for j := 0; j < 1<<(i+2); j++ {
			r.c[j] = r.c[j].Multiply(r.c[j])
		}
		ifft(r.c, i+2)
		for j := 1 << (i + 1); j < min(1<<(i+2), n+1); j++ {
			r.c[j] = r.c[j].Multiply(im)
		}
		if im.n&1 == 1 {
			im.n += mod
		}
		im.n /= 2
	}
	return r
}

func (p Poly) Sqrt(n int) Poly {
	if n == -1 {
		n = p.Degree()
	}
	if n == -1 {
		return p
	}
	if p.c[0].Equal(NewMint(1)) {
		if n == 0 {
			r := NewPoly(0)
			r.c[0] = NewMint(1)
			return r
		}
		k := 32 - clz(n)
		g := NewPoly((1 << (k - 1)) - 1)
		d := make([]Mint, 1<<k)
		e := make([]Mint, 1<<(k-1))
		f := make([]Mint, 1<<k)
		gg := make([]Mint, 1<<k)
		for i := 0; i < min(1<<(k-1), p.Degree()+1); i++ {
			g.c[i] = p.c[i]
		}
		g = g.Isqrt()
		for i := 0; i < 1<<(k-1); i++ {
			gg[i] = g.c[i]
		}
		fft(gg, k)
		for i := 0; i < 1<<k; i++ {
			f[i] = gg[i]
		}
		im := NewMint((mod + 1) / 2).Pow(k)
		for i := 0; i < 1<<k; i++ {
			gg[i] = gg[i].Multiply(f[i]).Multiply(f[i]).Multiply(f[i])
		}
		ifft(gg, k)
		for i := 1 << (k - 1); i < min(1<<k, n+1); i++ {
			g.c[i] = gg[i].Multiply(im)
		}
		return g
	}
	if p.c[0].Equal(NewMint(0)) {
		return p
	}
	n = p.Degree()
	k := 32 - clz(n)
	g := NewPoly((1 << (k - 1)) - 1)
	d := make([]Mint, 1<<(k+1))
	for i := 0; i < min(1<<(k-1), p.Degree()+1); i++ {
		g.c[i] = p.c[i]
	}
	for i := 1; i <= k; i++ {
		for j := 0; j < 1<<(k-i); j++ {
			t := (1 << (k - i)) + j
			g.c[t] = NewMint(1)
			for l := 0; l < 1<<i; l++ {
